packages/ipfs/src/core/index.js

```diff
@@ -133,3 +92,100 @@ module.exports = {
+ * in browsers)
+ * @property {object|Function} [libp2p] - The libp2p option allows you to build
+ * your libp2p node by configuration, or via a bundle function. If you are
+ * looking to just modify the below options, using the object format is the
+ * quickest way to get the default features of libp2p. If you need to create a
+ * more customized libp2p node, such as with custom transports or peer/content
+ * routers that need some of the ipfs data on startup, a custom bundle is a
+ * great way to achieve this.
+ * - You can see the bundle in action in the [custom libp2p example](https://github.com/ipfs/js-ipfs/tree/master/examples/custom-libp2p).
+ * - Please see [libp2p/docs/CONFIGURATION.md](https://github.com/libp2p/js-libp2p/blob/master/doc/CONFIGURATION.md)
+ *   for the list of options libp2p supports.
+ * - Default: [`libp2p-nodejs.js`](../src/core/runtime/libp2p-nodejs.js)
+ *   in Node.js, [`libp2p-browser.js`](../src/core/runtime/libp2p-browser.js) in
+ *   browsers.
+ */
+
+/**
+ * @typedef {IPFSRepo|string} RepoOption
+ * The file path at which to store the IPFS node’s data. Alternatively, you
+ * can set up a customized storage system by providing an `ipfs.Repo` instance.
+ *
+ * @example
+ * ```js
+ * // Store data outside your user directory
+ * const node = await IPFS.create({ repo: '/var/ipfs/data' })
+ * ```
+ * @typedef {import('./components/init').InitOptions} InitOptions
+ *
+ * @typedef {object} RelayOptions
+ * @property {boolean} [enabled] - Enable circuit relay dialer and listener. (Default: `true`)
+ * @property {object} [hop]
+ * @property {boolean} [hop.enabled] - Make this node a relay (other nodes can connect *through* it). (Default: `false`)
+ * @property {boolean} [hop.active] - Make this an *active* relay node. Active relay nodes will attempt to dial a destination peer even if that peer is not yet connected to the relay. (Default: `false`)
+ *
+ * @typedef {object} PreloadOptions
+ * @property {boolean} [enabled] - Enable content preloading (Default: `true`)
+ * @property {string[]} [addresses] - Multiaddr API addresses of nodes that should preload content.
+ *                                  - **NOTE:** nodes specified here should also be added to your node's bootstrap address list at `config.Boostrap`.
+ *
+ * @typedef {object} ExperimentalOptions
+ * @property {boolean} [ipnsPubsub] - Enable pub-sub on IPNS. (Default: `false`)
+ * @property {boolean} [sharding] - Enable directory sharding. Directories that have many child objects will be represented by multiple DAG nodes instead of just one. It can improve lookup performance when a directory has several thousand files or more. (Default: `false`)
+ *
+ * @typedef {import('ipfs-repo')} IPFSRepo
+ */
+
+/**
+ * Utility type to write type level conditionals
+ * @template Conditon, Then, Else
+ * @typedef {Conditon extends false
```

Comment:
```suggestion
 * @typedef {NonNullable<Conditon> extends false
```

If `"strictNullChecks"` is not set to `true` in `tsconifg.json`,
TypeScript cannot inference type properly unless `init: true, start: true` is explicitly specified

Reply:
This pull request does set it (see https://github.com/ipfs/js-ipfs/pull/3281/files#diff-9a28699e64db6125a0b49074aeb6fdd6R13)

Reply:
If `ipfs` is installed as a dependency, can `.d.ts` files use `tsconifg.json` here?

Reply:
@Xmader that is an interesting question and I get your concern now. I guess we could put back that `NonNullable<Condition>` to avoid that problem. That said, I think we should change `ipfs.create` to return API object that has same interface. To that end I have created #3285